### ISSUE TITLE
Question: list of mysql/mysql2 API incompatibilities

### ISSUE THREAD

**Issue by dtikhonov-iex** [sidorares/node-mysql2#1382]
The documentation states:

> MySQL2 is mostly API compatible with mysqljs and supports majority of features.

Is there an existing list of API incompatibilities between the two libraries?  It would make switching from mysql to mysql2 much easier.

**Comment by sidorares** [comment]
things that are known to behave differently are listed here: https://github.com/sidorares/node-mysql2/tree/master/documentation#known-incompatibilities-with-node-mysql ( hon sure how accurate is the list as of now, feel free to validate / extend, any help would be really appreciated! )

Also most apis in linked docs here https://github.com/sidorares/node-mysql2/tree/master/documentation#documentation are "extras" ( if you use it in your code with mysql2 you will have to modify to migrate back to mysql )

**Comment by dtikhonov-iex** [comment]
What I tripped over was that mysql's `query()` calls the callback with three arguments (the third argument, `fields`, always seems to be `undefined`), while mysql2's `query()` uses three arguments if `fields` is truthy, otherwise, two arguments are used:

```javascript
      if (fields) {
        process.nextTick(() => {
          this.onResult(null, rows, fields);
        });  
      } else {
        process.nextTick(() => {
          this.onResult(null, rows);
        });  
      }
```

This affects code that uses [denodeified](https://www.npmjs.com/package/q#adapting-node) version of the `query` function:  It receives two-element array argument `[rows, fields]` when using "mysql" or a single argument, `rows`, when using "mysql2".  It's not a big deal -- I modified my code to be able to handle either -- but it took a while to find.

(Yes, I love the "extras" -- that's why I plan on switching!)

**Comment by sidorares** [comment]
yeah, unfortunately easy to assume not passing optional argument as a backwards compatible, this caused some [similar pain in the past](https://github.com/sidorares/node-mysql2/pull/192)

**Comment by dtikhonov-iex** [comment]
I see.  Perhaps the fix, then, is to add this issue to the list of incompatibilities and move on.
